### ISSUE TITLE
config-custom.yaml override does not work in some cases

### ISSUE THREAD

**Issue by Zeromax** [puphpet/puphpet#1414]
Since #1370 the cool Feature is added to override some config stuff from the puphpet yaml.
This works like a charm, if the Option in the `config.yaml` exists.

like:
```yaml
vagrantfile-local:
    vm:
        box_url: 'special box url'
```
if the Option does not exist, the override does not work.
Original snippet `config.yaml`:
```yaml
mysql:
    install: '1'
    override_options: {  }
    root_password: root
    adminer: 0
    databases: {  }
```

the override `config-custom.yaml`:
```yaml
mysql:
    override_options:
        mysqld:
            datadir: '/vagrant/mysql-data'
```

**Comment by jtreminio** [comment]
config.yaml:

    ---
    vagrantfile-local:
        vm:
            box: puphpet/debian75-x64
            box_url: puphpet/debian75-x64
            hostname: ''
            memory: '512'
            cpus: '1'
            chosen_provider: virtualbox
            network:
                private_network: 192.168.56.176
                forwarded_port:
                    vK0tYUU2B7GB:
                        host: '7434'
                        guest: '22'
            post_up_message: ''
            provider:
                virtualbox:
                    modifyvm:
                        natdnshostresolver1: on
                vmware:
                    numvcpus: 1
                parallels:
                    cpus: 1
            provision:
                puppet:
                    manifests_path: puphpet/puppet
                    manifest_file: site.pp
                    module_path: puphpet/puppet/modules
                    options:
                        - '--verbose'
                        - '--hiera_config /vagrant/puphpet/puppet/hiera.yaml'
                        - '--parser future'
            synced_folder:
                0tZnqBUJsHuU:
                    owner: www-data
                    group: www-data
                    source: ./
                    target: /var/www
                    sync_type: default
                    rsync:
                        args:
                            - '--verbose'
                            - '--archive'
                            - '-z'
                        exclude:
                            - .vagrant/
                        auto: 'false'
            usable_port_range:
                start: 10200
                stop: 10500
        ssh:
            host: null
            port: null
            private_key_path: null
            username: vagrant
            guest_port: null
            keep_alive: true
            forward_agent: false
            forward_x11: false
            shell: 'bash -l'
        vagrant:
            host: detect
    server:
        install: '1'
        packages: {  }
    users_groups:
        install: '1'
        groups: {  }
        users: {  }
    cron:
        install: '1'
        jobs: {  }
    firewall:
        install: '1'
        rules: {  }
    apache:
        install: '1'
        settings:
            user: www-data
            group: www-data
            default_vhost: true
            manage_user: false
            manage_group: false
            sendfile: 0
        modules:
            - rewrite
        vhosts:
            8qvxyqfyhhj7:
                servername: awesome.dev
                serveraliases:
                    - www.awesome.dev
                docroot: /var/www/awesome
                port: '80'
                setenv:
                    - 'APP_ENV dev'
                directories:
                    tyvszaboc59w:
                        provider: directory
                        path: /var/www/awesome
                        options:
                            - Indexes
                            - FollowSymlinks
                            - MultiViews
                        allow_override:
                            - All
                        require:
                            - all
                            - granted
                        custom_fragment: ''
                engine: php
                custom_fragment: ''
                ssl_cert: ''
                ssl_key: ''
                ssl_chain: ''
                ssl_certs_dir: ''
        mod_pagespeed: 0
    nginx:
        install: '0'
        settings:
            default_vhost: 1
            proxy_buffer_size: 128k
            proxy_buffers: '4 256k'
        upstreams: {  }
        vhosts:
            sjwzrhjnjnfr:
                proxy: ''
                server_name: awesome.dev
                server_aliases:
                    - www.awesome.dev
                www_root: /var/www/awesome
                listen_port: '80'
                location: \.php$
                index_files:
                    - index.html
                    - index.htm
                    - index.php
                envvars:
                    - 'APP_ENV dev'
                engine: php
                client_max_body_size: 1m
                ssl_cert: ''
                ssl_key: ''
    php:
        install: '1'
        version: '56'
        composer: '1'
        composer_home: ''
        modules:
            php:
                - cli
                - intl
                - mcrypt
            pear: {  }
            pecl:
                - pecl_http
        ini:
            display_errors: On
            error_reporting: '-1'
            session.save_path: /var/lib/php/session
        timezone: America/Chicago
        mod_php: 0
    hhvm:
        install: '0'
        nightly: 0
        composer: '1'
        composer_home: ''
        settings:
            host: 127.0.0.1
            port: '9000'
        ini:
            display_errors: On
            error_reporting: '-1'
        timezone: null
    xdebug:
        install: '0'
        settings:
            xdebug.default_enable: '1'
            xdebug.remote_autostart: '0'
            xdebug.remote_connect_back: '1'
            xdebug.remote_enable: '1'
            xdebug.remote_handler: dbgp
            xdebug.remote_port: '9000'
    xhprof:
        install: '0'
    wpcli:
        install: '0'
        version: v0.17.1
    drush:
        install: '0'
        version: 6.3.0
    ruby:
        install: '1'
        versions: {  }
    nodejs:
        install: '0'
        npm_packages: {  }
    python:
        install: '1'
        packages: {  }
        versions: {  }
    mysql:
        install: '1'
        override_options: {  }
        root_password: '123'
        adminer: 0
        databases:
            T0kPg1dflOem:
                grant:
                    - ALL
                name: dbname
                host: localhost
                user: dbuser
                password: '123'
                sql_file: ''
    postgresql:
        install: '0'
        settings:
            root_password: '123'
            user_group: postgres
            encoding: UTF8
            version: '9.3'
        databases: {  }
        adminer: 0
    mariadb:
        install: '0'
        override_options: {  }
        root_password: '123'
        adminer: 0
        databases: {  }
        version: '10.0'
    sqlite:
        install: '0'
        adminer: 0
        databases: {  }
    mongodb:
        install: '0'
        settings:
            auth: 1
            port: '27017'
        databases: {  }
    redis:
        install: '0'
        settings:
            conf_port: '6379'
    mailcatcher:
        install: '0'
        settings:
            smtp_ip: 0.0.0.0
            smtp_port: 1025
            http_ip: 0.0.0.0
            http_port: '1080'
            mailcatcher_path: /usr/local/rvm/wrappers/default
            from_email_method: inline
    beanstalkd:
        install: '0'
        settings:
            listenaddress: 0.0.0.0
            listenport: '13000'
            maxjobsize: '65535'
            maxconnections: '1024'
            binlogdir: /var/lib/beanstalkd/binlog
            binlogfsync: null
            binlogsize: '10485760'
        beanstalk_console: 0
        binlogdir: /var/lib/beanstalkd/binlog
    rabbitmq:
        install: '0'
        settings:
            port: '5672'
    elastic_search:
        install: '0'
        settings:
            version: 1.4.1
            java_install: true
    solr:
        install: '0'
        settings:
            version: 4.10.2
            port: '8984'
    
config-custom.yaml:

    mysql:
        override_options:
            mysqld:
                datadir: '/vagrant/mysql-data'
    foo:
        bar: baz

output:

    {
      "vagrantfile-local"=>{
        "vm"=>{
          "box"=>"puphpet/debian75-x64",
          "box_url"=>"puphpet/debian75-x64",
          "hostname"=>"",
          "memory"=>"512",
          "cpus"=>"1",
          "chosen_provider"=>"virtualbox",
          "network"=>{
            "private_network"=>"192.168.56.176",
            "forwarded_port"=>{
              "vK0tYUU2B7GB"=>{
                "host"=>"7434",
                "guest"=>"22"
              }
            }
          },
          "post_up_message"=>"",
          "provider"=>{
            "virtualbox"=>{
              "modifyvm"=>{
                "natdnshostresolver1"=>true
              }
            },
            "vmware"=>{
              "numvcpus"=>1
            },
            "parallels"=>{
              "cpus"=>1
            }
          },
          "provision"=>{
            "puppet"=>{
              "manifests_path"=>"puphpet/puppet",
              "manifest_file"=>"site.pp",
              "module_path"=>"puphpet/puppet/modules",
              "options"=>[
                "--verbose",
                "--hiera_config /vagrant/puphpet/puppet/hiera.yaml",
                "--parser future"
              ]
            }
          },
          "synced_folder"=>{
            "0tZnqBUJsHuU"=>{
              "owner"=>"www-data",
              "group"=>"www-data",
              "source"=>"./",
              "target"=>"/var/www",
              "sync_type"=>"default",
              "rsync"=>{
                "args"=>[
                  "--verbose",
                  "--archive",
                  "-z"
                ],
                "exclude"=>[
                  ".vagrant/"
                ],
                "auto"=>"false"
              }
            }
          },
          "usable_port_range"=>{
            "start"=>10200,
            "stop"=>10500
          }
        },
        "ssh"=>{
          "host"=>nil,
          "port"=>nil,
          "private_key_path"=>nil,
          "username"=>"vagrant",
          "guest_port"=>nil,
          "keep_alive"=>true,
          "forward_agent"=>false,
          "forward_x11"=>false,
          "shell"=>"bash -l"
        },
        "vagrant"=>{
          "host"=>"detect"
        }
      },
      "server"=>{
        "install"=>"1",
        "packages"=>{
          
        }
      },
      "users_groups"=>{
        "install"=>"1",
        "groups"=>{
          
        },
        "users"=>{
          
        }
      },
      "cron"=>{
        "install"=>"1",
        "jobs"=>{
          
        }
      },
      "firewall"=>{
        "install"=>"1",
        "rules"=>{
          
        }
      },
      "apache"=>{
        "install"=>"1",
        "settings"=>{
          "user"=>"www-data",
          "group"=>"www-data",
          "default_vhost"=>true,
          "manage_user"=>false,
          "manage_group"=>false,
          "sendfile"=>0
        },
        "modules"=>[
          "rewrite"
        ],
        "vhosts"=>{
          "8qvxyqfyhhj7"=>{
            "servername"=>"awesome.dev",
            "serveraliases"=>[
              "www.awesome.dev"
            ],
            "docroot"=>"/var/www/awesome",
            "port"=>"80",
            "setenv"=>[
              "APP_ENV dev"
            ],
            "directories"=>{
              "tyvszaboc59w"=>{
                "provider"=>"directory",
                "path"=>"/var/www/awesome",
                "options"=>[
                  "Indexes",
                  "FollowSymlinks",
                  "MultiViews"
                ],
                "allow_override"=>[
                  "All"
                ],
                "require"=>[
                  "all",
                  "granted"
                ],
                "custom_fragment"=>""
              }
            },
            "engine"=>"php",
            "custom_fragment"=>"",
            "ssl_cert"=>"",
            "ssl_key"=>"",
            "ssl_chain"=>"",
            "ssl_certs_dir"=>""
          }
        },
        "mod_pagespeed"=>0
      },
      "nginx"=>{
        "install"=>"0",
        "settings"=>{
          "default_vhost"=>1,
          "proxy_buffer_size"=>"128k",
          "proxy_buffers"=>"4 256k"
        },
        "upstreams"=>{
          
        },
        "vhosts"=>{
          "sjwzrhjnjnfr"=>{
            "proxy"=>"",
            "server_name"=>"awesome.dev",
            "server_aliases"=>[
              "www.awesome.dev"
            ],
            "www_root"=>"/var/www/awesome",
            "listen_port"=>"80",
            "location"=>"\\.php$",
            "index_files"=>[
              "index.html",
              "index.htm",
              "index.php"
            ],
            "envvars"=>[
              "APP_ENV dev"
            ],
            "engine"=>"php",
            "client_max_body_size"=>"1m",
            "ssl_cert"=>"",
            "ssl_key"=>""
          }
        }
      },
      "php"=>{
        "install"=>"1",
        "version"=>"56",
        "composer"=>"1",
        "composer_home"=>"",
        "modules"=>{
          "php"=>[
            "cli",
            "intl",
            "mcrypt"
          ],
          "pear"=>{
            
          },
          "pecl"=>[
            "pecl_http"
          ]
        },
        "ini"=>{
          "display_errors"=>true,
          "error_reporting"=>"-1",
          "session.save_path"=>"/var/lib/php/session"
        },
        "timezone"=>"America/Chicago",
        "mod_php"=>0
      },
      "hhvm"=>{
        "install"=>"0",
        "nightly"=>0,
        "composer"=>"1",
        "composer_home"=>"",
        "settings"=>{
          "host"=>"127.0.0.1",
          "port"=>"9000"
        },
        "ini"=>{
          "display_errors"=>true,
          "error_reporting"=>"-1"
        },
        "timezone"=>nil
      },
      "xdebug"=>{
        "install"=>"0",
        "settings"=>{
          "xdebug.default_enable"=>"1",
          "xdebug.remote_autostart"=>"0",
          "xdebug.remote_connect_back"=>"1",
          "xdebug.remote_enable"=>"1",
          "xdebug.remote_handler"=>"dbgp",
          "xdebug.remote_port"=>"9000"
        }
      },
      "xhprof"=>{
        "install"=>"0"
      },
      "wpcli"=>{
        "install"=>"0",
        "version"=>"v0.17.1"
      },
      "drush"=>{
        "install"=>"0",
        "version"=>"6.3.0"
      },
      "ruby"=>{
        "install"=>"1",
        "versions"=>{
          
        }
      },
      "nodejs"=>{
        "install"=>"0",
        "npm_packages"=>{
          
        }
      },
      "python"=>{
        "install"=>"1",
        "packages"=>{
          
        },
        "versions"=>{
          
        }
      },
      "mysql"=>{
        "install"=>"1",
        "override_options"=>{
          "mysqld"=>{
            "datadir"=>"/vagrant/mysql-data"
          }
        },
        "root_password"=>"123",
        "adminer"=>0,
        "databases"=>{
          "T0kPg1dflOem"=>{
            "grant"=>[
              "ALL"
            ],
            "name"=>"dbname",
            "host"=>"localhost",
            "user"=>"dbuser",
            "password"=>"123",
            "sql_file"=>""
          }
        }
      },
      "postgresql"=>{
        "install"=>"0",
        "settings"=>{
          "root_password"=>"123",
          "user_group"=>"postgres",
          "encoding"=>"UTF8",
          "version"=>"9.3"
        },
        "databases"=>{
          
        },
        "adminer"=>0
      },
      "mariadb"=>{
        "install"=>"0",
        "override_options"=>{
          
        },
        "root_password"=>"123",
        "adminer"=>0,
        "databases"=>{
          
        },
        "version"=>"10.0"
      },
      "sqlite"=>{
        "install"=>"0",
        "adminer"=>0,
        "databases"=>{
          
        }
      },
      "mongodb"=>{
        "install"=>"0",
        "settings"=>{
          "auth"=>1,
          "port"=>"27017"
        },
        "databases"=>{
          
        }
      },
      "redis"=>{
        "install"=>"0",
        "settings"=>{
          "conf_port"=>"6379"
        }
      },
      "mailcatcher"=>{
        "install"=>"0",
        "settings"=>{
          "smtp_ip"=>"0.0.0.0",
          "smtp_port"=>1025,
          "http_ip"=>"0.0.0.0",
          "http_port"=>"1080",
          "mailcatcher_path"=>"/usr/local/rvm/wrappers/default",
          "from_email_method"=>"inline"
        }
      },
      "beanstalkd"=>{
        "install"=>"0",
        "settings"=>{
          "listenaddress"=>"0.0.0.0",
          "listenport"=>"13000",
          "maxjobsize"=>"65535",
          "maxconnections"=>"1024",
          "binlogdir"=>"/var/lib/beanstalkd/binlog",
          "binlogfsync"=>nil,
          "binlogsize"=>"10485760"
        },
        "beanstalk_console"=>0,
        "binlogdir"=>"/var/lib/beanstalkd/binlog"
      },
      "rabbitmq"=>{
        "install"=>"0",
        "settings"=>{
          "port"=>"5672"
        }
      },
      "elastic_search"=>{
        "install"=>"0",
        "settings"=>{
          "version"=>"1.4.1",
          "java_install"=>true
        }
      },
      "solr"=>{
        "install"=>"0",
        "settings"=>{
          "version"=>"4.10.2",
          "port"=>"8984"
        }
      },
      "foo"=>{
        "bar"=>"baz"
      }
    }

**Comment by Zeromax** [comment]
@jtreminio Thanks for your reoly.

It does not work either.
Steps:
1. vagrant destroy (if machine exists)
2. vagrant up

The mysql datadir is set to `/var/lib/mysql/`
I have added a extra mysql testart script like you said in https://github.com/puphpet/puphpet/issues/1303#issuecomment-78218254

If I add the Options to the original config.yaml file. everything works fine. So there must be a config load Problem somewhere.

For a further test, I execute a `vagrant provision` wich doesn't help to set the path correct. Set to `/var/lib/mysql/` again.

i tried also to Login via ssh and restart the mysql Server -> the path is set to `/var/lib/mysql/`

**Comment by marcosgdf** [comment]
Why closing #1303 if this is the same bug?

**Comment by Zeromax** [comment]
@marcosgdf it is not the same bug. They are related, but not the same. My Problem here occures only if I use the `config-custom.yaml`.
If I Change the lines in the `config.yaml` and restart the mysql Server as @jtreminio proposed in https://github.com/puphpet/puphpet/issues/1303#issuecomment-78218254 everything works fine.

so #1303 is a different bug...

**Comment by marcosgdf** [comment]
OK. Sorry :\

**Comment by Zeromax** [comment]
@marcosgdf No Problem it is a bit confusing ;)

**Comment by jtreminio** [comment]
@Zeromax I've been working on this for ~8 hours now. I'm going to mark this as won't fix unless someone else comes in with some help.

The current puppet-mysql module puphpet uses, 2.3.1, seems to have a bug where it doesn't actually set datadir. It passes the default value when starting MySQL, ignoring whatever's been set in config.yaml

Upgrade to latest version 3.3.0, has another bug. Provisioning fails because the command `mysql_install_db --basedir=/usr --defaults-extra-file=/etc/mysql/my.cnf --datadir=/vagrant/mysql-data --user=mysql` errors with `FATAL ERROR: Could not find my-default.cnf`. The "solution" is to create an empty `/usr/my-default.cnf` file. However, more bugs pop up. The last one I saw and attempted to debug was `No directory, logging in with HOME=` at which point I gave up.

Upgrading to 3.1.0 experiences same issue.

I'm going to merge in the ability to define your own settings for both MySQL and MariaDB (which will introduce a breaking change in the config.yaml that you'll need to note), but I will not offer support when it breaks (it probably will, from what I'm seeing).
